### PR TITLE
Pending Audit Logs are sent in batches

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

- Part of #9870 
  - I'm not closing the ticket, as there is still some more work to be done. If the logging is faster than the background thread can keep up and the log queue grows out of control, we still need to implement some backpressure mechanism to make even async logs to block for a bit to give the background thread time to process the backlog.
  - This ticket only implements one crucial optimization - batching. If there are multiple logs in the queue, they are sent in a single request (up to 100 messages) instead of doing a separate request for each log message.
  - This does not delay sending the logs. Each log is tried to be sent as long as the background thread is ready to process it - we are **not** waiting for multiple log messages to gather them to form a bigger batch.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.

### CODE REVIEW

**Comment by radeusgd** [comment]
I also manually tested on real cloud to ensure logging still works in practice:
```
> Duration.time_execution <| 0.up_to 100 . each x-> Audit_Log.report_event "TestEvent" "FOO-RADEK-TEST-"+x.to_text async=(x != 99)
>>> (Pair.Value PT0.886909S Nothing)
```
image

Sending these 100 messages took about 1s. If instead I do `async=False` meaning that I'm waiting for each message to be sent before sending another one, this takes 9.5s. So batching seems to work there as well.

File: std-bits/base/src/main/java/org/enso/base/enso_cloud/audit/LogJobsQueue.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package org.enso.base.enso_cloud.audit;
+
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+
+class LogJobsQueue {
```
Comment (JaroslavTulach) [inline std-bits/base/src/main/java/org/enso/base/enso_cloud/audit/LogJobsQueue.java:8]: Nice, isolated monitor.

File: std-bits/base/src/main/java/org/enso/base/enso_cloud/audit/AuditLogMessage.java [hunk 2]
```diff
@@ -8,7 +8,7 @@
 import org.enso.base.CurrentEnsoProject;
 import org.enso.base.enso_cloud.CloudAPI;
 
-public class AuditLogMessage implements AuditLogAPI.LogMessage {
+public class AuditLogMessage implements AuditLogApiAccess.LogMessage {
```
Comment (JaroslavTulach) [inline std-bits/base/src/main/java/org/enso/base/enso_cloud/audit/AuditLogMessage.java]: A public class extends non-public class!? It is not really nice to expose such internals in an API.

File: std-bits/base/src/main/java/org/enso/base/enso_cloud/audit/AuditLogApiAccess.java [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+  }
+
+  public void logWithoutConfirmation(LogMessage message) {
+    enqueueJob(new LogJob(message, null));
+  }
+
+  private void enqueueJob(LogJob job) {
+    ensureConfigSaved();
+    int queuedJobs = logQueue.enqueue(job);
+    if (queuedJobs == 1 && backgroundThreadService.getQueue().isEmpty()) {
```
Comment (JaroslavTulach) [inline std-bits/base/src/main/java/org/enso/base/enso_cloud/audit/AuditLogApiAccess.java:61]: I wasn't sure about this `queuedJobs == 1` check, but when I realized `logQueue.enqueue` is synchronized method, I believe it is going to work properly.

File: std-bits/base/src/main/java/org/enso/base/enso_cloud/audit/AuditLogApiAccess.java [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+  }
+
+  /** Runs as long as there are any pending log messages queued and sends them in batches. */
+  private void logThreadEntryPoint() {
+    while (true) {
+      List<LogJob> pendingMessages = logQueue.popEnqueuedJobs(MAX_BATCH_SIZE);
+      if (pendingMessages.isEmpty()) {
+        // If there are no more pending messages, we can stop the thread for now.
+        // If during this teardown a new message is added, it will see no elements on `logQueue` and thus,
+        // `logQueue.enqueue` will return 1, thus ensuring that at least one new job is scheduled.
```
Comment (JaroslavTulach) [inline std-bits/base/src/main/java/org/enso/base/enso_cloud/audit/AuditLogApiAccess.java:89]: +1

**Review by JaroslavTulach — APPROVED** [review]
I'd say the executor's finishing/starting logic is correct.

File: std-bits/base/src/main/java/org/enso/base/enso_cloud/audit/AuditLogMessage.java (hunk 2 above)
Comment (radeusgd) [reply std-bits/base/src/main/java/org/enso/base/enso_cloud/audit/AuditLogMessage.java]: Hmm, you are right, `AuditLogMessage` does not need to and probably should never have been public, I will make it package-private.

**Review by GregoryTravis — APPROVED** [review]

**Review by AdRiley — APPROVED** [review]

**Review by jdunkerley — APPROVED** [review]

**Pull request merged**